cleanup(libsinsp): consolidation and extension of libsinsp stats / metrics `sinsp_stats_v2 `

incertum opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines in the [CONTRIBUTING.md](https://github.com/falcosecurity/.github/blob/master/CONTRIBUTING.md) file and learn how to compile Falco from source [here](https://falco.org/docs/source).
2. Please label this pull request according to what type of issue you are addressing.
3. Please add a release note!
4. If the PR is unfinished while opening it specify a wip in the title before the actual title, for example, "wip: my awesome feature"
-->

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

/kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

<!--
Please remove the leading whitespace before the `/kind <>` you uncommented.
-->

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area API-version

> /area build

> /area CI

> /area driver-kmod

> /area driver-bpf

> /area driver-modern-bpf

> /area libscap-engine-bpf

> /area libscap-engine-gvisor

> /area libscap-engine-kmod

> /area libscap-engine-modern-bpf

> /area libscap-engine-nodriver

> /area libscap-engine-noop

> /area libscap-engine-source-plugin

> /area libscap-engine-savefile

> /area libscap-engine-udig

> /area libscap

> /area libpman

/area libsinsp

> /area tests

> /area proposals

<!--
Please remove the leading whitespace before the `/area <>` you uncommented.
-->

**Does this PR require a change in the driver versions?**

> /version driver-API-version-major

> /version driver-API-version-minor

> /version driver-API-version-patch

> /version driver-SCHEMA-version-major

> /version driver-SCHEMA-version-minor

> /version driver-SCHEMA-version-patch

<!--
Please remove the leading whitespace before the `/version <>` you uncommented.
-->

**What this PR does / why we need it**:

See https://github.com/falcosecurity/libs/issues/1347: Expanding libsinsp stats will help to debug and optimize libsinsp functionality around the threadtable and other aspects in production settings.

In essence, this PR refactors / consolidates / simplifies the libsinsp stats interface and adds valuable metrics that already existed, but weren't easily accessible by clients such as Falco. Plus added overall CPU and memory usages of the host to put the agents' usage easier in relation to the current server load.

**Justification for removing the old sinsp stats:**

First every metric was preserved with the replacement solution, except the counters below were duplicative of the scap counters, hence we do not re-copy them in the new sinsp stats solution

```
m_stats->m_n_seen_evts = stats.n_evts;
m_stats->m_n_drops = stats.n_drops;
m_stats->m_n_preemptions = stats.n_preemptions;
```

The following 2 metrics are snapshot metrics anyways, therefore they now take place in `libsinsp::stats::get_sinsp_stats_v2` just like the resource utilization snapshots.

```
uint64_t m_n_threads;
uint64_t m_n_fds;
```

As said all other counters that were on the hot path are preserved in the simple `sinsp_stats_v2` struct.

Please note that `GATHER_INTERNAL_STATS` is currently broken in 2 places: When activated libsinsp would not compile because of missing updates in `fdinfo.h` and a legacy threadtable loop approach `void sinsp_thread_manager::update_statistics()` -> This is why I believe no one could have seriously used those stats for some time.

**Additional information:**

During the October core maintainer meeting we discussed, that we are all fine with having all these stats / metrics in the `sinsp_stats_v2_buffer` and while this PR at least removes and consolidate legacy libsinsp stats, we are all on the same page that as we further cleanup and refactor libsinsp the libsinsp stats functionality will be part of future cleanups.

**Question:**

Are we ok with just having all these counters just like that on the hot path or do we want to gate them? Meaning another if condition or is it cheaper to just increment or decrement the respective count real fast?

**Which issue(s) this PR fixes**:

https://github.com/falcosecurity/libs/issues/1347

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
If PR is `kind/failing-tests` please post the related issues/tests in a comment and do not use `Fixes`.
-->

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

<!--
If no, you have to do nothing.
If yes, a release note is required:
Delete `NONE` and enter your extended release note in the block below.
Please note, the release note follows the "conventional commit specification" (https://www.conventionalcommits.org/en/v1.0.0/):
For example: `fix: broken link`.
If the PR requires additional action from users switching to the new release, prepend the string "action required:".
For example, `action required: change the API interface of libscap`.
-->

```release-note
cleanup(libsinsp): consolidation and extension of libsinsp stats / metrics `sinsp_stats_v2 `
```

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/1433#" title="Author self-approved">incertum</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [incertum]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

incertum commented:
The corresponding Falco PR is up and here is some example output of the new metrics https://github.com/falcosecurity/falco/pull/2883#issuecomment-1773518211

incertum commented:
Had a new idea on how to expose additional stats / metrics around the container engine cache.

Missing container images in the logs has been called out for a very very long time. While in some cases when a new container just comes up we can't do much about it, we at least want to make sure everything we can improve has been considered.

It will help to investigate https://github.com/falcosecurity/falco/issues/2708 better.

FedeDP commented on userspace/libsinsp/sinsp.h:
```diff
@@ -1121,7 +1122,8 @@ VISIBILITY_PRIVATE
 	const scap_machine_info* m_machine_info;
 	const scap_agent_info* m_agent_info;
-	scap_stats_v2 m_sinsp_stats_v2[SINSP_MAX_RESOURCE_UTILIZATION];
+	sinsp_stats_v2 m_sinsp_stats_v2;
```
Since stats were protected by a macro before, can we retains a similar approach by making this back a `unique_ptr` and only creating it if requested (ie: by a `sinsp_enable_stats()` API or similar that must be called before the `open`ing of the inspector).

gnosek replied:
Maybe we don't need a unique_ptr, just ensure the struct is zero-initialized and have an option to disable the stats updates?

If you disable the stats and then read them and expect non-zero values, I'm not sure you have someone else to blame :D

FedeDP replied:
This would save us lots of `if m_sinsp_stats_v2 != nil` checks, neat!

incertum replied:
@gnosek ^ you also onboard, if yes can go ahead and re-push later.

incertum replied:
@FedeDP pushed a suggestion, plz let me know. It seems checking the inspector would only be needed in one place and in fact in most cases it's already de-referenced anyways before.

In addition, luckily had a pod sandbox running in the background and realized we should not be counting sandboxes as missing container images ... that's the one important metric I really would like to check on in prod, so we better get it right.

FedeDP reviewed (commented):
I like this PR; generally speaking, i'd avoid the check for `m_inspector != nullptr` since that is already taken for granted in so many places.

Also, i'd disable stats by default and i'd add a `sinsp_enable_stats` to enable them, making the `m_sinsp_stats_v2` field a unique_ptr that is only allocated if stats are enabled.

incertum commented:
@FedeDP thanks for answering the question if we want a flag to enable everything.

Re the null checks, some unit tests create conditions where the inspector in fact would be NULL, hence follow up question:
- Change unit tests?
- Check if inspector is null only sometimes?
- Never check if inspector is null?

Since you didn't object against removing the old implementation I assume that means it's ok?

gnosek commented on userspace/libsinsp/fdinfo.cpp:
```diff
@@ -345,6 +345,49 @@ sinsp_fdtable::sinsp_fdtable(sinsp* inspector)
 	reset_cache();
 }
 
+sinsp_fdinfo_t* sinsp_fdtable::find(int64_t fd)
```
in that case I won't complain about using `auto` instead of the explicit iterator type and the atrocious indentation :D

incertum commented:
@FedeDP and @gnosek I added the flag in for now and kept is as simple struct as before as @gnosek suggested, but I can always change it if needed. Personally feel these stats should be as simple as possible in any regards.

Also cleaned up the git history a little bit.

Once we decide on https://github.com/falcosecurity/libs/pull/1433#issuecomment-1779652266 I'll make those changes.

incertum commented on userspace/libsinsp/test/sinsp_with_test_input.h:
```diff
@@ -59,7 +59,8 @@ class sinsp_with_test_input : public ::testing::Test {
 	void open_inspector(scap_mode_t mode = SCAP_MODE_TEST) {
-		m_inspector.open_test_input(m_test_data.get(), mode);
+		m_inspector.open_test_input(m_test_data.get(), mode);	
+		m_inspector.set_sinsp_stats_v2_enabled(true);
```
Note: Setting this in the unit test only did not work.

incertum commented on userspace/libsinsp/sinsp.cpp:
```diff
@@ -2708,6 +2660,27 @@ void sinsp::set_proc_scan_log_interval_ms(uint64_t val)
 
+void sinsp::set_sinsp_stats_v2_enabled()
+{
+	m_sinsp_stats_v2 = std::make_unique<sinsp_stats_v2>();
```
@FedeDP forgot to add checks if it's NULL in case you would call it multiple times. Will re-push, but I'll wait for your feedback first in case there is more.

incertum commented:
@jasondellaluce is also on board with this interim refactor that will allow us to consume the new metrics.
I created https://github.com/falcosecurity/libs/issues/1463 to track future API refactors we already agreed on.

One more question: Do we or can we merge this and the Falco PR this week? Else I would propose the week of Nov 13th as Jason and I will be out at KubeCon NA.

This Falco PR https://github.com/falcosecurity/falco/pull/2879 also needs to be merged before the corresponding Falco PR https://github.com/falcosecurity/falco/pull/2883 or I consolidate those 2 PRs in one.
In any case we should update Falco so it's not incompatible anymore with current libs master.

Andreagit97 commented on userspace/libsinsp/threadinfo.cpp:
```diff
@@ -1444,14 +1444,6 @@ void sinsp_thread_manager::clear()
 	m_last_tid = 0;
 	m_last_flush_time_ns = 0;
 	m_n_drops = 0;
-
```
SInce we have removed all these fields at this point i would remove also `m_n_drops` and add it to sinsp_stats. Maybe we could use a more meaningful name like `full_threadtable_drops`. In this way we can handle this stat like all the others without a dedicated method `get_m_n_drops`

incertum replied:
Done, but came at a small costs in terms of changing a g_logger log that now only fires if sinsp stats v2 are enabled.

Andreagit97 commented on userspace/libsinsp/threadinfo.cpp:
```diff
@@ -1537,9 +1529,10 @@ std::unique_ptr<sinsp_threadinfo> sinsp_thread_manager::new_threadinfo() const
-#ifdef GATHER_INTERNAL_STATS
-	m_added_threads->increment();
-#endif
+	if (m_inspector->m_sinsp_stats_v2)
+	{
+		m_inspector->m_sinsp_stats_v2->m_n_added_threads++;
+	}
```
I would move this to the end of the function because here it is not counting the `added_threads`, the table could be full. This counts the `requests` of adding a thread

Andreagit97 reviewed (commented):
I left some comments. To be honest, I'm not fully convinced by the approach we are following with these stats, we have sinsp_stats, machine_info_stats, agent_info_stats, and resource_stats ... some in the platforms some with a shared pointer in the inspector and then populated with an external method `libsinsp::stats::get_...`. Probably before going on with further steps, I would like to plan a final design to understand what is the final aim because at the moment it is not clear to me

Andreagit97 commented on userspace/libsinsp/stats.cpp:
```diff
@@ -16,64 +16,406 @@ limitations under the License.
+		buffer[SINSP_STATS_V2_N_FDS].value.u64 = 0;
+		threadinfo_map_t* threadtable = thread_manager->get_threads();
+		threadtable->loop([&] (sinsp_threadinfo& tinfo) {
+			sinsp_fdtable* fdtable = tinfo.get_fd_table();
```
here we need a check `sinsp_fdtable != nullptr`

incertum replied:
:heart: more segfaults is the last thing we need, which makes we wonder if we could come up with a more proper project inspection in this regard as sometimes we say we trust the pointer to be never NULL and prefer to save the overhead of checking vs sometimes we definitely need to check ...

Andreagit97 commented:
Before going on with code changes, I have some general questions not related to the code:
* I saw that in Falco (and I imagine also in other consumers) we use the `get_capture_stats` method to obtain the number of drops/events. On the other hand, with `stats_v2` we are using an agnostic approach, where the final consumer receives a vector of metrics already populated by sinsp. My question here is, do we want `scap_stats_v2` to replace the old `scap_stats`? If yes, how do we obtain the specific number of drops/events from this agnostic approach? Do we want to keep these specific numbers or the final goal is to expose a set of metrics with a Prometheus endpoint?
* If the final goal is to expose only a Prometheus endpoint using something like https://github.com/jupp0r/prometheus-cpp, are we doing the right thing by exposing a vector of `struct scap_stats_v2`? At this point we could directly create Prometheus structs in sinsp and expose them to the consumers, so they can add them to the registry and expose them.

WDYT?

incertum commented on userspace/libsinsp/sinsp.cpp:
```diff
@@ -2792,6 +2744,30 @@ void sinsp::set_proc_scan_log_interval_ms(uint64_t val)
+{
+	if (m_sinsp_stats_v2 == nullptr)
+	{
+		m_sinsp_stats_v2 = std::make_unique<sinsp_stats_v2>();
```
correct needs to be a shared pointer exactly because of what you wrote here

incertum commented on userspace/libsinsp/stats.h:
```diff
@@ -17,45 +17,80 @@ limitations under the License.
+	uint64_t m_n_failed_thread_lookups;
+	uint64_t m_n_added_threads;
+	uint64_t m_n_removed_threads;
+	uint32_t m_n_missing_container_images;
```
Seems large enough and uint64_t likely never needed. I doubt machines would go above 200-500 containers per host at a time anytime soon ... and let me know if we need additional comments to clarify that these container counters are not monotonic and the current view only similar to n_threads and n_fds. We could make them monotonic, but then we need to hook into the container engine differently. I know it's always a mess and believe me likely no winning ever with finding the perfect metrics.

incertum commented on userspace/libsinsp/threadinfo.cpp:
```diff
@@ -1721,9 +1714,11 @@ void sinsp_thread_manager::remove_thread(int64_t tid)
-		m_failed_lookups->increment();
-#endif
+		// Extra m_inspector nullptr check
+		if (m_inspector != nullptr && m_inspector->m_sinsp_stats_v2)
```
Yes the problem child condition: asan will fail if we don't check here because of `TEST(sinsp_thread_manager, remove_non_existing_thread)`, change the unit test?

incertum commented:
@Andreagit97 thank you, I have moved the discussion to relevant issues.

Andreagit97 reviewed (approved):
/approve

poiana commented:
LGTM label has been added.  <details>Git tree hash: b1e038d84fffa650a145d188fae3ef2244f92de1</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/1433#pullrequestreview-1737013029" title="Approved">Andreagit97</a>*, *<a href="https://github.com/falcosecurity/libs/pull/1433#" title="Author self-approved">incertum</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [Andreagit97,incertum]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

jasondellaluce reviewed (approved):
/approve

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/1433#pullrequestreview-1737013029" title="Approved">Andreagit97</a>*, *<a href="https://github.com/falcosecurity/libs/pull/1433#" title="Author self-approved">incertum</a>*, *<a href="https://github.com/falcosecurity/libs/pull/1433#pullrequestreview-1737682760" title="Approved">jasondellaluce</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [Andreagit97,incertum,jasondellaluce]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->
